src/main/java/org/ambraproject/wombat/controller/ArticleController.java

```diff
@@ -258,70 +258,98 @@ public XmlContent(String html, List<Reference> references) {
+   *
+   * @param site
+   * @param articlePointer
+   * @param xml
+   * @param references
+   * @return
+   * @throws IOException
+   */
+  private String getArticleHtml(Site site, ArticlePointer articlePointer, byte[] xml, List<Reference> references) throws IOException {
+    StringWriter articleHtml = new StringWriter(XFORM_BUFFER_SIZE);
+    try (OutputStream outputStream = new WriterOutputStream(articleHtml, charset)) {
+      articleTransformService.transformArticle(site, articlePointer, references, new ByteArrayInputStream(xml), outputStream);
+    }
+    return articleHtml.toString();
+  }
 
-      StringWriter articleHtml = new StringWriter(XFORM_BUFFER_SIZE);
-      try (OutputStream outputStream = new WriterOutputStream(articleHtml, charset)) {
-        articleTransformService.transformArticle(site, articlePointer, references,
-            new ByteArrayInputStream(xml), outputStream);
+  /**
+   * Get a list of references.
+   * References to PLOS articles are supplemented with the absolute link to the journal site.
+   *
+   * @param site
+   * @param request
+   * @param xml
+   * @return
+   * @throws IOException
+   */
+  private List<Reference> getReferences(Site site, HttpServletRequest request, byte[] xml) throws IOException {
+    final Document document = parseXmlService.getDocument(new ByteArrayInputStream(xml));
+    // do not supply Solr related link service now
+    List<Reference> references = parseXmlService.parseArticleReferences(document, null);
+
+    // invoke the Solr API once to resolve all journal keys
+    List<String> plosDois = references.stream().map(ref -> ref.getDoi()).filter(doi -> isPlosJournal(doi)).collect(Collectors.toList());
+    List<String> plosJournalKeys = doiToJournalResolutionService.getJournalKeysFromDois(plosDois, site);
+
+    // store the link text from journal key to references.
+    // since Reference is immutable, need to create a new list of new reference objects.
+    Iterator<Reference> refs = references.iterator();
+    Iterator<String> journalKeys = plosJournalKeys.iterator();
+    List<Reference> processedReferences = new ArrayList<Reference>();
+    while (refs.hasNext()) {
+      Reference ref = refs.next();
+      if (!isPlosJournal(ref.getDoi())) {
+        processedReferences.add(ref);
+        continue;
       }
 
-      return new XmlContent(articleHtml.toString(), references);
-    });
+      String key = journalKeys.next();
+      if (Strings.isNullOrEmpty(key)) {
+        processedReferences.add(ref);
+        continue;
+      }
+
+      Reference.Builder builder = new Reference.Builder(ref);
+      Reference refWithLink = builder.setFullArticleLink(getFullArticleLink(site, request, ref.getDoi(), key)).build();
+      processedReferences.add(refWithLink);
+    }
+
+    references = processedReferences;
+    return references;
   }
 
-  private Boolean inPlosJournal(String doi) {
+  private Boolean isPlosJournal(String doi) {
```

Comment:
I think the old name was more clear - it checks if the `doi` is of an article published *in* a PLOS journal.